irohad/multi_sig_transactions/state/impl/mst_state.cpp

```diff
@@ -172,92 +187,110 @@ namespace iroha {
   }
 
   MstState::MstState(const CompleterType &completer,
-                     size_t transaction_limit,
+                     std::shared_ptr<BatchStorageLimit> storage_limit,
                      logger::LoggerPtr log)
       : MstState(completer,
-                 transaction_limit,
+                 std::move(storage_limit),
                  std::vector<DataType>{},
                  std::move(log)) {}
 
   MstState::MstState(const CompleterType &completer,
-                     size_t transaction_limit,
+                     std::shared_ptr<BatchStorageLimit> storage_limit,
                      const BatchesForwardCollectionType &batches,
                      logger::LoggerPtr log)
       : completer_(completer),
-        txs_limit_(transaction_limit),
+        batches_(std::move(storage_limit), std::make_unique<InternalStorage>()),
         txs_quantity_(0),
         log_(std::move(log)) {
     for (const auto &batch : batches) {
-      batches_.insert({oldestTimestamp(batch), batch});
+      batches_.insert(batch);
       txs_quantity_ += batch->transactions().size();
     }
   }
 
   void MstState::insertOne(StateUpdateResult &state_update,
                            const DataType &rhs_batch) {
     log_->info("batch: {}", *rhs_batch);
-    auto corresponding = batches_.right.find(rhs_batch);
-    if (corresponding == batches_.right.end()) {
-      // when state does not contain transaction
-      if (transactionsQuantity() + boost::size(rhs_batch->transactions())
-          <= txs_limit_) {
-        // there is enough room for the new batch
-        rawInsert(rhs_batch);
-        state_update.updated_state_->rawInsert(rhs_batch);
-      } else {
-        // there is not enough room for the new batch
-        log_->info(
-            "Dropped a batch because it would exceed the transaction limit "
-            "(currently have {} out of {} transactions in state): {}",
-            transactionsQuantity(),
-            txs_limit_,
-            *rhs_batch);
+    auto completed_batches = batches_.move([this, &state_update, &rhs_batch](
+                                               auto &storage)
+                                               -> std::vector<BatchPtr> {
+      auto corresponding = storage.batches.right.find(rhs_batch);
```

Comment:
I would be very glad if you will use any other variable name instead of "corresponding". Hope this is possible.